qt-models/treemodel.h

```diff
@@ -25,16 +26,16 @@ class TreeModel : public QAbstractItemModel {
 	Q_OBJECT
 public:
 	TreeModel(QObject *parent = 0);
-	~TreeModel();
 	QVariant data(const QModelIndex &index, int role) const;
 	int rowCount(const QModelIndex &parent = QModelIndex()) const;
 	int columnCount(const QModelIndex &parent = QModelIndex()) const;
 	QModelIndex index(int row, int column, const QModelIndex &parent = QModelIndex()) const;
 	QModelIndex parent(const QModelIndex &child) const;
+	void clear();
 
 protected:
 	int columns;
-	TreeItem *rootItem;
```

Comment:
QScopedPointer is the Qt alternative here.

i still don't know if we have a policy on C++ stdlib vs Qt.
like, when to use what? is there are a reason for using one over the other? etc.

my thinking would be - use Qt unless stdlib is better at something.

Reply:
Here, `QScopedPointer` would work. But in general, I had such a bad experience with `QScopedPointer` that this is the one case where I would generally go with `std::unique_ptr`. The difference is that the latter has move-semantics, which means that you can pass down ownership. For example you can put `std::unique_ptr`s into (stdlib) containers. See also https://stackoverflow.com/questions/40346393/should-i-use-qscopedpointer-or-stdunique-ptr

For `QVector` vs `std::vector`, etc. I think we should decide on a case-by-case basis. I tend to default to the former, even though I much prefer the latter. They are surprisingly different: Qt does reference counting and copy-on-write, whereas stdlib is just syntactic sugar on top of C `malloc`/`realloc`. You pass it by value and it will make a deep copy (which is what passing by value means). Qt devs argue that their containers are more "beginner"-friendly, because you don't have to think about ownership, but I have my doubts because this leads to quite unpredictable behavior. Personally, I think Qt's containers are broken, but that it's not worth fighting them.

Also Qt-versions have convenience functions such as `vector.contains(element)`, vs. in stdlib you would write a single general function for that: `contains(general_container, element)`. The second is much closer to C, no?

Reply:
thank you for the explanation.
the overall move to use std::* more widely in the codebase was a bit of a surprise, mostly because nobody investigated differences between the standard library and Qt containers before.

if one tries to contribute stdlib code in the Qt source code, it would be rejected for consistency reasons - i.e. use the Qt containers.
so from a consistency standing point this is not exactly good for Subsurface, but i guess we can do it here and there as long as the code is readable.

we should keep in mind that the C++ standard library tends to not be very readable.

Reply:
Yes, with a bit of a heavy heart, I try to use Qt-containers unless COW semantics make it impossible or distinctly more difficult. An example is putting owning pointers in containers. Suppose you had a `QVector<std::unique_ptr>`. Modifying means possible copying and - boom - you would suddenly have two versions of the unique-ptr pointing to the same object, which is by definition not possible. In my opinion this just shows that Qt containers are broken with respect to ownership management, but what can you do. Instead of fixing their containers they decided to invent QML. :(